File: chap2/PreCG1.f90
!// 在病态矩阵问题上，共轭梯度法比部分主元的高斯消去法还要差
!// 这种情况下，可通过预条件得到缓解，主要是将问题转化为良态矩阵系统，再实施CG方法
!// 此代码的预条件因子M=D，为雅可比预条件因子，D为A的对角线矩阵
Program ConjugateGradientMethods
  Implicit none
  Integer :: i, j
  Integer, parameter :: n = 3
  Real(kind=8) :: A(n,n) = [ 4.d0, -2.d0, 2.d0, -2.d0, 2.d0, -4.d0, 2.d0, -4.d0, 11.d0 ]
  Real(kind=8) :: b(n,1) = [ 6.d0, -10.d0, 27.d0 ]
  Real(kind=8) :: alpha = 0.d0, beta = 0.d0
  Real(kind=8) :: x(n,1) = 0.d0, d(n,1) = 0.d0, r(n,1) = 0.d0, z(n,1) = 0.d0
  Real(kind=8) :: rtmp(n,1) = 0.d0, tmp(1,1) = 0.d0, ztmp(n,1) = 0.d0
  Real(kind=8) :: M(n,n) = 0.d0
  
  forall ( i = 1:n, j = 1:n, i == j ) M(i,j) = A(i,j)
  forall ( i = 1:n, j = 1:n, i == j ) M(i,j) = 1.d0 / M(i,j)
  
  r = b - matmul( A,x ) !// r0 = b - Ax0
  z = matmul( M,r )  !// z0 = Inv(M) * r0
  d = z  !// d0 = z0

  Do i = 0, n - 1
    If ( all(r<0.0) ) then
      exit
    End if
    rtmp = r
    ztmp = z
    tmp = matmul( transpose(rtmp),ztmp ) / matmul( matmul( transpose(d),A ), d )
    alpha = tmp(1,1)
    x = x + alpha * d
    r = r - alpha * matmul( A,d )
    z = matmul( M,r )
    tmp = matmul( transpose(r),z ) / matmul( transpose(rtmp),ztmp )
    beta = tmp(1,1)
    d = z + beta * d
  End do
  
  Do i = 1, n
    Write ( *,'(f12.5)' ) x(i,1)
  End do
End program ConjugateGradientMethods

File: chap2/PreCG1.py
# -*- coding: utf-8 -*-
"""
Created on Fri May 18 10:57:51 2018

@author: luk
"""

"""
在病态矩阵问题上，共轭梯度法比部分主元的高斯消去法还要差
这种情况下，可通过预条件得到缓解，主要是将问题转化为良态矩阵系统，再实施CG方法
此代码的预条件因子M=D，为雅可比预条件因子，D为A的对角线矩阵
"""
import numpy as np
n = 3
a = np.array( [[4., -2., 2. ],[-2., 2., -4.],[2., -4., 11.]], dtype = np.float64 )
b = np.array( [6., -10., 27.], dtype = np.float64 )
alpha = 0.0
beta = 0.0
M = np.zeros_like( a, dtype = np.float64 )
x = np.zeros( (n,), dtype = np.float64 )
d = np.zeros_like( x, dtype = np.float64 )
r = np.zeros_like( x, dtype = np.float64 )
z = np.zeros_like( x, dtype = np.float64 )
rtmp = np.zeros_like( x, dtype = np.float64 )
ztmp = np.zeros_like( x, dtype = np.float64 )

for i in range(n):
    M[i,i] = a[i,i]
    M[i,i] = 1.0 / M[i,i]
    
r = b - np.matmul( a,x )  # r0 = b - Ax0
z = np.matmul( M,r )  # z0 = Inv(M)*r0
d = z  # d0 = z0

for i in range(n):
    if ( all(r<0.0) ):
        quit()
    
    rtmp = r
    ztmp = z
    tmp = np.matmul( np.transpose(rtmp),ztmp ) / np.matmul( np.matmul( np.transpose(d),a ), d )
    alpha = tmp
    x = x + alpha * d
    r = r - alpha * np.matmul( a,d )
    z = np.matmul( M,r )
    tmp = np.matmul( np.transpose(r),z ) / np.matmul( np.transpose(rtmp),ztmp )
    beta = tmp
    d = z + beta * d
    
print( " the x of PreCG1 is :" )
for i in range(n):
    print( x[i] )
    
print( "the x of python is :" )  
a = np.array( [[4., -2., 2. ],[-2., 2., -4.],[2., -4., 11.]], dtype = np.float64 )
b = np.array( [6., -10., 27.], dtype = np.float64 )
a = np.linalg.inv(a)  # 求矩阵a的逆矩阵
x = np.matmul(a,b)
for i in range(len(x)):
      print( x[i] )
      
print( " please input Enter and stop!" )
input()